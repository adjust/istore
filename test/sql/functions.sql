SELECT exist('1=>1'::istore, 1);
SELECT exist('1=>1'::istore, 2);
SELECT exist('1=>1, -1=>0'::istore, 2);
SELECT exist('1=>1, -1=>0', -1);

SELECT fetchval('1=>1'::istore, 1);
SELECT fetchval('2=>1'::istore, 1);
SELECT fetchval('1=>1, 1=>1'::istore, 1);
SELECT fetchval('1=>1, 1=>1'::istore, 2);

SELECT add('1=>1, 2=>1'::istore, '1=>1, 2=>1'::istore);
SELECT add('1=>1, 2=>1'::istore, '-1=>1, 2=>1'::istore);
SELECT add('1=>1, 2=>1'::istore, '-1=>-1, 2=>1'::istore);
SELECT add('-1=>1, 2=>1'::istore, '-1=>-1, 2=>1'::istore);
SELECT add('-1=>-1, 2=>1'::istore, '-1=>-1, 2=>1'::istore);
SELECT add('-1=>-1, 2=>1'::istore, 1);
SELECT add('-1=>-1, 2=>1'::istore, -1);
SELECT add('-1=>-1, 2=>1'::istore, 0);

SELECT subtract('1=>1, 2=>1'::istore, '1=>1, 2=>1'::istore);
SELECT subtract('1=>1, 2=>1'::istore, '-1=>1, 2=>1'::istore);
SELECT subtract('1=>1, 2=>1'::istore, '-1=>-1, 2=>1'::istore);
SELECT subtract('-1=>1, 2=>1'::istore, '-1=>-1, 2=>1'::istore);
SELECT subtract('-1=>-1, 2=>1'::istore, '-1=>-1, 2=>1'::istore);
SELECT subtract('-1=>-1, 2=>1'::istore, 1);
SELECT subtract('-1=>-1, 2=>1'::istore, -1);
SELECT subtract('-1=>-1, 2=>1'::istore, 0);

SELECT multiply('1=>1, 2=>1'::istore, '1=>1, 2=>1'::istore);
SELECT multiply('1=>1, 2=>1'::istore, '-1=>1, 2=>1'::istore);
SELECT multiply('1=>1, 2=>1'::istore, '-1=>-1, 2=>1'::istore);
SELECT multiply('-1=>1, 2=>1'::istore, '-1=>-1, 2=>1'::istore);
SELECT multiply('-1=>-1, 2=>1'::istore, '-1=>-1, 2=>1'::istore);
SELECT multiply('-1=>-1, 2=>1'::istore, 1);
SELECT multiply('-1=>-1, 2=>1'::istore, -1);
SELECT multiply('-1=>-1, 2=>1'::istore, 0);

SELECT istore_from_array(ARRAY[1]);
SELECT istore_from_array(ARRAY[1,1,1,1]);
SELECT istore_from_array(NULL);
SELECT istore_from_array(ARRAY[1,2,3,4]);
SELECT istore_from_array(ARRAY[1,2,3,4,1,2,3,4]);
SELECT istore_from_array(ARRAY[1,2,3,4,1,2,3,NULL]);
SELECT istore_from_array(ARRAY[NULL,2,3,4,1,2,3,4]);
SELECT istore_from_array(ARRAY[NULL,2,3,4,1,2,3,NULL]);
SELECT istore_from_array(ARRAY[1,2,3,NULL,1,NULL,3,4,1,2,3]);

SELECT istore_agg(ARRAY['1=>1']::istore[]);
SELECT istore_agg(ARRAY['1=>1','1=>1']::istore[]);
SELECT istore_agg(ARRAY['1=>1,2=>1','1=>1,2=>-1']::istore[]);
SELECT istore_agg(ARRAY['1=>1,2=>1','1=>1,2=>-1',NULL]::istore[]);
SELECT istore_agg(ARRAY[NULL,'1=>1,2=>1','1=>1,2=>-1']::istore[]);
SELECT istore_agg(ARRAY[NULL,'1=>1,2=>1','1=>1,2=>-1',NULL]::istore[]);

SELECT istore_sum_up('1=>1'::istore);
SELECT istore_sum_up(NULL::istore);
SELECT istore_sum_up('1=>1, 2=>1'::istore);
SELECT istore_sum_up('1=>1 ,2=>-1, 1=>1'::istore);

BEGIN;
CREATE TABLE test (a istore);
INSERT INTO test VALUES('1=>1');
INSERT INTO test VALUES('2=>1');
INSERT INTO test VALUES('3=>1');
SELECT SUM(a) FROM test;
ROLLBACK;

BEGIN;
CREATE TABLE test (a istore);
INSERT INTO test VALUES('1=>1');
INSERT INTO test VALUES('2=>1');
INSERT INTO test VALUES('3=>1');
INSERT INTO test VALUES(NULL);
INSERT INTO test VALUES('3=>3');
SELECT SUM(a) FROM test;
ROLLBACK;
