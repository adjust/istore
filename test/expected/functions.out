SELECT exist('1=>1'::istore, 1);
 exist 
-------
 t
(1 row)

SELECT exist('1=>1'::istore, 2);
 exist 
-------
 f
(1 row)

SELECT exist('1=>1, -1=>0'::istore, 2);
 exist 
-------
 f
(1 row)

SELECT exist('1=>1, -1=>0', -1);
 exist 
-------
 t
(1 row)

SELECT fetchval('1=>1'::istore, 1);
 fetchval 
----------
        1
(1 row)

SELECT fetchval('2=>1'::istore, 1);
 fetchval 
----------
         
(1 row)

SELECT fetchval('1=>1, 1=>1'::istore, 1);
 fetchval 
----------
        2
(1 row)

SELECT fetchval('1=>1, 1=>1'::istore, 2);
 fetchval 
----------
         
(1 row)

SELECT add('1=>1, 2=>1'::istore, '1=>1, 2=>1'::istore);
        add        
-------------------
 "1"=>"2","2"=>"2"
(1 row)

SELECT add('1=>1, 2=>1'::istore, '-1=>1, 2=>1'::istore);
             add             
-----------------------------
 "-1"=>"1","1"=>"1","2"=>"2"
(1 row)

SELECT add('1=>1, 2=>1'::istore, '-1=>-1, 2=>1'::istore);
             add              
------------------------------
 "-1"=>"-1","1"=>"1","2"=>"2"
(1 row)

SELECT add('-1=>1, 2=>1'::istore, '-1=>-1, 2=>1'::istore);
        add         
--------------------
 "-1"=>"0","2"=>"2"
(1 row)

SELECT add('-1=>-1, 2=>1'::istore, '-1=>-1, 2=>1'::istore);
         add         
---------------------
 "-1"=>"-2","2"=>"2"
(1 row)

SELECT add('-1=>-1, 2=>1'::istore, 1);
        add         
--------------------
 "-1"=>"0","2"=>"2"
(1 row)

SELECT add('-1=>-1, 2=>1'::istore, -1);
         add         
---------------------
 "-1"=>"-2","2"=>"0"
(1 row)

SELECT add('-1=>-1, 2=>1'::istore, 0);
         add         
---------------------
 "-1"=>"-1","2"=>"1"
(1 row)

SELECT subtract('1=>1, 2=>1'::istore, '1=>1, 2=>1'::istore);
     subtract      
-------------------
 "1"=>"0","2"=>"0"
(1 row)

SELECT subtract('1=>1, 2=>1'::istore, '-1=>1, 2=>1'::istore);
           subtract           
------------------------------
 "-1"=>"-1","1"=>"1","2"=>"0"
(1 row)

SELECT subtract('1=>1, 2=>1'::istore, '-1=>-1, 2=>1'::istore);
          subtract           
-----------------------------
 "-1"=>"1","1"=>"1","2"=>"0"
(1 row)

SELECT subtract('-1=>1, 2=>1'::istore, '-1=>-1, 2=>1'::istore);
      subtract      
--------------------
 "-1"=>"2","2"=>"0"
(1 row)

SELECT subtract('-1=>-1, 2=>1'::istore, '-1=>-1, 2=>1'::istore);
      subtract      
--------------------
 "-1"=>"0","2"=>"0"
(1 row)

SELECT subtract('-1=>-1, 2=>1'::istore, 1);
      subtract       
---------------------
 "-1"=>"-2","2"=>"0"
(1 row)

SELECT subtract('-1=>-1, 2=>1'::istore, -1);
      subtract      
--------------------
 "-1"=>"0","2"=>"2"
(1 row)

SELECT subtract('-1=>-1, 2=>1'::istore, 0);
      subtract       
---------------------
 "-1"=>"-1","2"=>"1"
(1 row)

SELECT multiply('1=>1, 2=>1'::istore, '1=>1, 2=>1'::istore);
     multiply      
-------------------
 "1"=>"1","2"=>"1"
(1 row)

SELECT multiply('1=>1, 2=>1'::istore, '-1=>1, 2=>1'::istore);
          multiply           
-----------------------------
 "-1"=>"0","1"=>"0","2"=>"1"
(1 row)

SELECT multiply('1=>1, 2=>1'::istore, '-1=>-1, 2=>1'::istore);
          multiply           
-----------------------------
 "-1"=>"0","1"=>"0","2"=>"1"
(1 row)

SELECT multiply('-1=>1, 2=>1'::istore, '-1=>-1, 2=>1'::istore);
      multiply       
---------------------
 "-1"=>"-1","2"=>"1"
(1 row)

SELECT multiply('-1=>-1, 2=>1'::istore, '-1=>-1, 2=>1'::istore);
      multiply      
--------------------
 "-1"=>"1","2"=>"1"
(1 row)

SELECT multiply('-1=>-1, 2=>1'::istore, 1);
      multiply       
---------------------
 "-1"=>"-1","2"=>"1"
(1 row)

SELECT multiply('-1=>-1, 2=>1'::istore, -1);
      multiply       
---------------------
 "-1"=>"1","2"=>"-1"
(1 row)

SELECT multiply('-1=>-1, 2=>1'::istore, 0);
      multiply      
--------------------
 "-1"=>"0","2"=>"0"
(1 row)

SELECT istore_from_array(ARRAY[1]);
 istore_from_array 
-------------------
 "1"=>"1"
(1 row)

SELECT istore_from_array(ARRAY[1,1,1,1]);
 istore_from_array 
-------------------
 "1"=>"4"
(1 row)

SELECT istore_from_array(NULL);
 istore_from_array 
-------------------
 
(1 row)

SELECT istore_from_array(ARRAY[1,2,3,4]);
          istore_from_array          
-------------------------------------
 "1"=>"1","2"=>"1","3"=>"1","4"=>"1"
(1 row)

SELECT istore_from_array(ARRAY[1,2,3,4,1,2,3,4]);
          istore_from_array          
-------------------------------------
 "1"=>"2","2"=>"2","3"=>"2","4"=>"2"
(1 row)

SELECT istore_from_array(ARRAY[1,2,3,4,1,2,3,NULL]);
          istore_from_array          
-------------------------------------
 "1"=>"2","2"=>"2","3"=>"2","4"=>"1"
(1 row)

SELECT istore_from_array(ARRAY[NULL,2,3,4,1,2,3,4]);
          istore_from_array          
-------------------------------------
 "1"=>"1","2"=>"2","3"=>"2","4"=>"2"
(1 row)

SELECT istore_from_array(ARRAY[NULL,2,3,4,1,2,3,NULL]);
          istore_from_array          
-------------------------------------
 "1"=>"1","2"=>"2","3"=>"2","4"=>"1"
(1 row)

SELECT istore_from_array(ARRAY[1,2,3,NULL,1,NULL,3,4,1,2,3]);
          istore_from_array          
-------------------------------------
 "1"=>"3","2"=>"2","3"=>"3","4"=>"1"
(1 row)

SELECT istore_agg(ARRAY['1=>1']::istore[]);
 istore_agg 
------------
 "1"=>"1"
(1 row)

SELECT istore_agg(ARRAY['1=>1','1=>1']::istore[]);
 istore_agg 
------------
 "1"=>"2"
(1 row)

SELECT istore_agg(ARRAY['1=>1,2=>1','1=>1,2=>-1']::istore[]);
    istore_agg     
-------------------
 "1"=>"2","2"=>"0"
(1 row)

SELECT istore_agg(ARRAY['1=>1,2=>1','1=>1,2=>-1',NULL]::istore[]);
    istore_agg     
-------------------
 "1"=>"2","2"=>"0"
(1 row)

SELECT istore_agg(ARRAY[NULL,'1=>1,2=>1','1=>1,2=>-1']::istore[]);
    istore_agg     
-------------------
 "1"=>"2","2"=>"0"
(1 row)

SELECT istore_agg(ARRAY[NULL,'1=>1,2=>1','1=>1,2=>-1',NULL]::istore[]);
    istore_agg     
-------------------
 "1"=>"2","2"=>"0"
(1 row)

SELECT istore_sum_up('1=>1'::istore);
 istore_sum_up 
---------------
             1
(1 row)

SELECT istore_sum_up(NULL::istore);
 istore_sum_up 
---------------
              
(1 row)

SELECT istore_sum_up('1=>1, 2=>1'::istore);
 istore_sum_up 
---------------
             2
(1 row)

SELECT istore_sum_up('1=>1 ,2=>-1, 1=>1'::istore);
 istore_sum_up 
---------------
             1
(1 row)

BEGIN;
CREATE TABLE test (a istore);
INSERT INTO test VALUES('1=>1');
INSERT INTO test VALUES('2=>1');
INSERT INTO test VALUES('3=>1');
SELECT SUM(a) FROM test;
            sum             
----------------------------
 "1"=>"1","2"=>"1","3"=>"1"
(1 row)

ROLLBACK;
BEGIN;
CREATE TABLE test (a istore);
INSERT INTO test VALUES('1=>1');
INSERT INTO test VALUES('2=>1');
INSERT INTO test VALUES('3=>1');
INSERT INTO test VALUES(NULL);
INSERT INTO test VALUES('3=>3');
SELECT SUM(a) FROM test;
            sum             
----------------------------
 "1"=>"1","2"=>"1","3"=>"4"
(1 row)

ROLLBACK;
